File: voice-router/menu_flow.go
package main

import (
	"context"
	"log/slog"
	"strings"
	"time"
)

func playMainMenu(callControlID string) {
	sendCommand(callControlID, "actions/gather_using_speak", map[string]interface{}{
		"language":                   "en-US",
		"voice":                      cfg.TelnyxVoice,
		"payload":                    cfg.IVRMainMenuText,
		"minimum_digits":             1,
		"maximum_digits":             1,
		"valid_digits":               "01",
		"timeout_millis":             8000,
		"inter_digit_timeout_millis": 5000,
	})
}

func handleGatherEnded(p CallPayload) {
	state, ok := getCall(p.CallControlID)
	if !ok {
		slog.Warn("Gather ended for unknown call", "call_id", p.CallControlID)
		return
	}

	digits := p.Digits
	slog.Info("Caller pressed digit", "digit", digits, "call_id", p.CallControlID)

	switch digits {
	case "1":
		// Transfer to LiveKit AI Support Agent
		if cfg.LiveKitSIPURI == "" {
			slog.Warn("LIVEKIT_SIP_URI not configured — falling back to voicemail")
			startVoicemail(p.CallControlID)
			return
		}
		state.PendingLiveKitTransfer = true
		say(p.CallControlID, cfg.IVRTransferText)

	case "0":
		// Voicemail
		startVoicemail(p.CallControlID)

	default:
		// No input or invalid — retry menu
		state.MenuRetries++
		if state.MenuRetries >= 2 {
			slog.Info("Max menu retries reached — routing to voicemail", "call_id", p.CallControlID)
			startVoicemail(p.CallControlID)
			return
		}
		playMainMenu(p.CallControlID)
	}
}

func handleSpeakEnded(p CallPayload) {
	state, ok := getCall(p.CallControlID)
	if !ok {
		return
	}

	// Generic transfer requested after the most recent speak completes.
	if to := strings.TrimSpace(state.PendingTransferTo); to != "" {
		state.PendingTransferTo = ""
		slog.Info("Transferring call after speak", "call_id", p.CallControlID, "to", to)
		sendCommand(p.CallControlID, "actions/transfer", map[string]interface{}{"to": to})
		return
	}

	// Voice agent: mark agent as done speaking, flush buffered user input.
	if cfg.EnableVoiceAgent {
		onAgentSpeakEnded(p.CallControlID)
	}

	if state.InVoicemail || cfg.LiveKitSIPURI == "" || !state.PendingLiveKitTransfer || state.LiveKitTransferred {
		return
	}
	state.PendingLiveKitTransfer = false
	state.LiveKitTransferred = true
	state.Status = "waiting"
	state.LiveKitRoom = ""

	sipTo := strings.TrimSpace(cfg.LiveKitSIPURI)
	if sipTo == "" {
		slog.Warn("LIVEKIT_SIP_URI not configured — falling back to voicemail")
		startVoicemail(p.CallControlID)
		return
	}

	slog.Info("Transferring call to LiveKit SIP", "sip", sipTo)
	sendCommand(p.CallControlID, "actions/transfer", map[string]interface{}{
		"to": sipTo,
	})

	// Start escalation timer once caller is in the waiting state.
	scheduleEscalation(p.CallControlID)

	// Start hold music only after the SIP participant is actually in the LiveKit room
	// created by the SIP dispatch rule.
	go func(callControlID, caller string) {
		// If the call already hung up, bail.
		if _, ok := getCall(callControlID); !ok {
			return
		}

		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		roomName, err := waitForDispatchedLiveKitRoom(ctx, caller, sipUserPart(cfg.LiveKitSIPURI))
		if err != nil {
			slog.Error("LiveKit SIP transfer did not result in a dispatched room", "call_id", callControlID, "from", caller, "err", err, "sip_user", sipUserPart(cfg.LiveKitSIPURI))
			return
		}

		state, ok := getCall(callControlID)
		if !ok {
			return
		}
		state.LiveKitRoom = roomName

		holdRoomsMu.Lock()
		if _, exists := holdRooms[callControlID]; exists {
			holdRoomsMu.Unlock()
			return
		}
		holdRoomsMu.Unlock()

		bridge, bErr := NewHoldRoomSystem(roomName, callControlID, cfg)
		if bErr != nil {
			slog.Error("Failed to start hold music in dispatched room", "call_id", callControlID, "room", roomName, "err", bErr)
			return
		}

		holdRoomsMu.Lock()
		holdRooms[callControlID] = bridge
		holdRoomsMu.Unlock()
		go bridge.Start()
	}(p.CallControlID, state.From)
}

File: telephony/src/lib/menuFlow.ts
// telephony/src/lib/menuFlow.ts
// Converted from menu_flow.go

import { sendCommand, say } from "./telnyxClient";
import { getConfig } from "./config";
import { getCall } from "./state";
import { startVoicemail } from "./voicemail";
import { scheduleEscalation } from "./escalation";
import { HoldRoomSystem, holdRooms } from "./holdRoomSystem";
import { waitForDispatchedLiveKitRoom, liveKitRoomName, sipUserPart } from "./livekitRoom";
import { onAgentSpeakEnded } from "./virtualAgent";
import type { CallPayload } from "../types/telnyx";

export function playMainMenu(callControlID: string): void {
  const cfg = getConfig();
  sendCommand(callControlID, "actions/gather_using_speak", {
    language: "en-US",
    voice: cfg.telnyxVoice,
    payload: cfg.ivrMainMenuText,
    minimum_digits: 1,
    maximum_digits: 1,
    valid_digits: "01",
    timeout_millis: 8000,
    inter_digit_timeout_millis: 5000,
  });
}

export function handleGatherEnded(p: CallPayload): void {
  const cfg = getConfig();
  const state = getCall(p.call_control_id);
  if (!state) {
    console.warn("[menu] Gather ended for unknown call", { call_id: p.call_control_id });
    return;
  }

  const digits = p.digits ?? "";
  console.log("[menu] Caller pressed digit", { digit: digits, call_id: p.call_control_id });

  switch (digits) {
    case "1":
      if (!cfg.liveKitSIPURI) {
        console.warn("[menu] LIVEKIT_SIP_URI not configured — falling back to voicemail");
        startVoicemail(p.call_control_id);
        return;
      }
      state.pendingLiveKitTransfer = true;
      say(p.call_control_id, cfg.ivrTransferText);
      break;

    case "0":
      startVoicemail(p.call_control_id);
      break;

    default:
      state.menuRetries = (state.menuRetries ?? 0) + 1;
      if ((state.menuRetries ?? 0) >= 2) {
        console.log("[menu] Max retries — routing to voicemail", { call_id: p.call_control_id });
        startVoicemail(p.call_control_id);
        return;
      }
      playMainMenu(p.call_control_id);
  }
}

export function handleSpeakEnded(p: CallPayload): void {
  const cfg = getConfig();
  const state = getCall(p.call_control_id);
  if (!state) return;

  // Generic transfer after speak
  if (state.pendingTransferTo?.trim()) {
    const to = state.pendingTransferTo;
    state.pendingTransferTo = "";
    console.log("[menu] Transferring after speak", { call_id: p.call_control_id, to });
    sendCommand(p.call_control_id, "actions/transfer", { to });
    return;
  }

  // Voice agent: flush buffered utterances
  if (cfg.enableVoiceAgent) {
    onAgentSpeakEnded(p.call_control_id);
  }

  if (
    state.inVoicemail ||
    !cfg.liveKitSIPURI ||
    !state.pendingLiveKitTransfer ||
    state.liveKitTransferred
  ) {
    return;
  }

  state.pendingLiveKitTransfer = false;
  state.liveKitTransferred = true;
  state.status = "waiting";
  state.livekit_room = "";

  const sipTo = cfg.liveKitSIPURI.trim();
  if (!sipTo) {
    console.warn("[menu] LIVEKIT_SIP_URI not configured — falling back to voicemail");
    startVoicemail(p.call_control_id);
    return;
  }

  console.log("[menu] Transferring to LiveKit SIP", { sip: sipTo });
  sendCommand(p.call_control_id, "actions/transfer", { to: sipTo });

  scheduleEscalation(p.call_control_id);

  // Start hold room monitoring once caller is in LiveKit
  const callControlID = p.call_control_id;
  const callerFrom = state.from;

  void (async () => {
    const st = getCall(callControlID);
    if (!st) return;

    try {
      const ac = new AbortController();
      const roomName = await waitForDispatchedLiveKitRoom(
        ac.signal,
        callerFrom,
        sipUserPart(cfg.liveKitSIPURI)
      );

      const s2 = getCall(callControlID);
      if (!s2) return;
      s2.livekit_room = roomName;

      if (holdRooms.has(callControlID)) return;

      const bridge = new HoldRoomSystem(roomName, callControlID);
      holdRooms.set(callControlID, bridge);
      bridge.start();
    } catch (err) {
      console.error("[menu] LiveKit SIP room not found", { call_id: callControlID, err });
    }
  })();
}
